File: java/timesheet/src/test/java/timesheet/EmployeeBuilder.java
package timesheet;

import java.time.LocalDate;

public class EmployeeBuilder {

    private String firstName = "John";
    private String lastName = "Doe";
    private LocalDate dateOfBirth = LocalDate.of(1980, 9, 10);
    private String email = "john.doe@foobar.com";
    private Timesheet timesheet = new Timesheet(0);

    public static EmployeeBuilder anEmployee(){
        return new EmployeeBuilder();
    }

    public EmployeeBuilder withEmail(String email) {
        this.email = email;
        return this;
    }

    public EmployeeBuilder withFirstName(String firstName) {
        this.firstName = firstName;
        return this;
    }


    public EmployeeBuilder withLastName(String lastName) {
        this.lastName = lastName;
        return this;
    }

    public EmployeeBuilder withDateOfBirth(LocalDate dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
        return this;
    }
    public EmployeeBuilder lastTimesheetAsOf(int monthNumber) {
        this.timesheet = new Timesheet(monthNumber);
        return this;
    }

    public Employee build(){
        return new Employee(firstName, lastName, dateOfBirth, email, timesheet);
    }

}

File: scala/timesheet/src/test/scala/timesheet/EmployeeBuilder.scala
package timesheet

import java.time.LocalDate

class EmployeeBuilder {
  private var firstName: String = "John"
  private var lastName: String = "Doe"
  private var dateOfBirth: LocalDate = LocalDate.of(1980, 9, 10)
  private var email: String = "john.doe@foobar.com"
  private var timesheet: Timesheet = Timesheet(0)

  def withEmail(email: String): EmployeeBuilder = {
    this.email = email
    this
  }

  def withFirstName(firstName: String): EmployeeBuilder = {
    this.firstName = firstName
    this
  }

  def withLastName(lastName: String): EmployeeBuilder = {
    this.lastName = lastName
    this
  }

  def withDateOfBirth(dateOfBirth: LocalDate): EmployeeBuilder = {
    this.dateOfBirth = dateOfBirth
    this
  }

  def lastTimesheetAsOf(monthNumber: Int): EmployeeBuilder = {
    this.timesheet = Timesheet(monthNumber)
    this
  }

  def build(): Worker = {
    new Employee(firstName, lastName, dateOfBirth, email, Some(timesheet))
  }
}

object EmployeeBuilder {
  def anEmployee(): EmployeeBuilder = new EmployeeBuilder()
}
